### PR TITLE
[YUNIKORN-1957] Add user/group limit config change e2e test

### PR DESCRIPTION
**Pull request by pegasas**
### What is this PR for?

use the unit test cases in yunikorn-core for add e2e test

### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [x] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?

[[YUNIKORN-1957] Add user/group limit config change e2e test](https://issues.apache.org/jira/browse/YUNIKORN-1957)

### How should this be tested?

Specific user/group limits limits changes to ensure order of precedence:

5. set user limit & wild card user limit only and ensure it has been honoured.
5a) When the user limit is specified, it should be considered for that specific user
5b) When the user limit is changed, make sure it meets new limit of for that specific group
6. set group limit & wild card group limit only and ensure it has been honoured.
6a) When the group limit is specified, it should be considered for that specific group
6b) When the group limit is changed, make sure it meets new limit of for that specific group
7. Repeat the above for max resources and max applications individually and together as well

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by pbacsko** [comment]
@pegasas please resolve the conflict to proceed

**Comment by pegasas** [comment]
> @pegasas please resolve the conflict to proceed

Hi, @pbacsko , conflicts are resolved.
Also, I am referencing the sample involved in https://github.com/apache/yunikorn-core/commit/0a05068035b71f1a810fad7c38f8cb0eb7af58f6, it seems not to work.

**Comment by doupache** [comment]
Hi @pegasas, thanks for the effort!  I am working on https://issues.apache.org/jira/browse/YUNIKORN-1950
But i found a bug and try to fix it. 

add to following test to the end TestUserGroupLimitChange test. 

```go 
//remove Limits , so we can run TestApp2 again
conf.Queues[0].Queues[0].Limits = nil
assert.NilError(t, manager.UpdateConfig(conf.Queues[0], "root"))

increased = manager.IncreaseTrackedResource(queuePathParent, TestApp2, usage, tc.user)
assert.Equal(t, increased, true, "unable to increase tracked resource: queuepath "+queuePathParent+", app "+TestApp2+", res "+usage.String())

decreased = manager.DecreaseTrackedResource(queuePathParent, TestApp2, usage, tc.user, true)
assert.Equal(t, decreased, true, "unable to decreased tracked resource: queuepath "+queuePathParent+", app "+TestApp2+", res "+usage.String())

//set newLimits again, we should not run TestApp2 now
conf.Queues[0].Queues[0].Limits = tc.newLimits
assert.NilError(t, manager.UpdateConfig(conf.Queues[0], "root"))

increased = manager.IncreaseTrackedResource(queuePathParent, TestApp2, usage, tc.user)
assert.Equal(t, increased, false, "should not increase tracked resource: queuepath "+queuePathParent+", app "+TestApp2+", res "+usage.String())
```

**Comment by pbacsko** [comment]
@manirajv06, you have more context regarding the feature. Please check if the comments above.

**Comment by manirajv06** [comment]
> > @pegasas please resolve the conflict to proceed
> 
> Hi, @pbacsko , thank you for your quick reply. conflicts are resolved. Also, I am referencing the sample involved in [apache/yunikorn-core@0a05068](https://github.com/apache/yunikorn-core/commit/0a05068035b71f1a810fad7c38f8cb0eb7af58f6), it seems not to work as expected.
> 
> It is my case not in right way or there is bug with wildcard?

I don't think above mentioned core commit should affect this test behaviour. Can you resolve the conflict and try running again?

**Comment by pegasas** [comment]
> I don't think above mentioned core commit should affect this test behaviour. Can you resolve the conflict and try running again?

Thanks @manirajv06 

It seems conflicts was introduced by https://github.com/apache/yunikorn-k8shim/commit/dd02f2e9055e1c262295a5aa87e9072a6f9d9c83 which was merged yesterday

 I have fixed this and re-commit.

File: test/e2e/user_group_limit/user_group_limit_test.go [hunk 1]
```diff
@@ -460,24 +491,14 @@ var _ = ginkgo.Describe("UserGroupLimit", func() {
...
-
-		// usergroup2 can't deploy the second sleep pod to root.sandbox1
-		deploySleepPod(usergroup2, "root.sandbox1", false, "because final application count is more than wildcard maxapplications")
-		checkUsage(userTestType, user2, "root.sandbox1", []*v1.Pod{usergroup2Sandbox1Pod1})
+		usergroup1Sandbox1Pod1 = deploySleepPod(usergroup1, "root.sandbox1", true, "because memory usage is less than maxresources")
+		checkUsage(userTestType, user1, "root.sandbox1", []*v1.Pod{usergroup1Sandbox1Pod1})
 	})
 
-	ginkgo.It("Verify_maxresources_with_a_wildcard_group_limit", func() {
+	ginkgo.It("Verify_maxapplications_with_users_limits_changed", func() {
```
[the remark this answers is not in the document]
Comment (manirajv06) [reply test/e2e/user_group_limit/user_group_limit_test.go:632]: It seems this block is for users as I can see 2 blocks below for groups. If it is true, then we need to check the config in this block as it doesn't have any limits for users

**Review by manirajv06 — CHANGES_REQUESTED** [review]

**Review by pbacsko — CHANGES_REQUESTED** [review]
@pegasas please investigate&fix the test failures:

```
Summarizing 4 Failures:
  [FAIL] UserGroupLimit [It] Verify_maxresources_with_users_limits_changed
  /home/runner/work/yunikorn-k8shim/yunikorn-k8shim/test/e2e/user_group_limit/user_group_limit_test.go:841
  [FAIL] UserGroupLimit [It] Verify_maxapplications_with_users_limits_changed
  /home/runner/work/yunikorn-k8shim/yunikorn-k8shim/test/e2e/user_group_limit/user_group_limit_test.go:885
  [FAIL] UserGroupLimit [It] Verify_maxresources_with_groups_limits_changed
  /home/runner/work/yunikorn-k8shim/yunikorn-k8shim/test/e2e/user_group_limit/user_group_limit_test.go:841
  [FAIL] UserGroupLimit [It] Verify_maxapplications_with_groups_limits_changed
  /home/runner/work/yunikorn-k8shim/yunikorn-k8shim/test/e2e/user_group_limit/user_group_limit_test.go:841
```

**Comment by pbacsko** [comment]
ping @pegasas have you had time to check the test failures?

**Pull request closed without merging**